lib/parsers/package-lock-parser.ts

```diff
@@ -49,66 +64,240 @@ export class PackageLockParser implements LockfileParser {
-      if (!lockfilePath.length) {
-        throw new Error(`Dependency ${depName} was not found in package-lock.json.
-          Your package.json and package-lock.json are probably out of sync.
-          Please run "npm install" and try again.`);
+      depPath.pop();
+    }
+
+    if (!depMap[depName]) {
+      throw new Error(`Dependency ${depName} was not found in package-lock.json.
+        Your package.json and package-lock.json are probably out of sync.
+        Please run "npm install" and try again.`);
+    }
+
+    return depName;
+  }
+
+  // Algorithm is based on dynamic programming technique and tries to build
+  // "more simple" trees and compose them into bigger ones.
+  private async createDepTrees(depMap: DepMap, depGraph): Promise<{[depPath: string]: PkgTree}> {
+
+    function setImmediatePromise() {
+      return new Promise((resolve) => setImmediate(resolve));
+    }
+
+    // Graph has to be acyclic
+    if (!graphlib.alg.isAcyclic(depGraph)) {
+      throw new Error('Cycles were not removed from graph.');
+    }
+
+    const depTrees: {[depPath: string]: PkgTree} = {};
+    // topological sort guarantees that all the dependencies, which doesn't require
+    // anything else, will be processed first
+    const depOrder = graphlib.alg.topsort(depGraph);
+
+    while (depOrder.length) {
+      const depKey = depOrder.shift() as string;
+      const dep = depMap[depKey];
+
+      // direction is from the dependency to the package requiring it, so we are
+      // looking for predecessors
+      for (const subDepPath of depGraph.predecessors(depKey)) {
+        const subDep = depTrees[subDepPath];
+        dep.dependencies[subDep.name] = subDep;
       }
-      // dependency was not found on a current path, remove last key (move closer to the root) and try again
-      // visitedDepPaths can be passed by a reference, because traversing up doesn't update it
-      return this.buildSubTreeRecursiveFromPackageLock(depName, lockfilePath.slice(0, -1), lockFile, depPath, isDev);
+      delete dep.requires;
+      depTrees[depKey] = {...dep as PkgTree};
+      // Since this code doesn't handle any I/O or network, we need to force
+      // event loop to tick while being used in server for request processing
+      await setImmediatePromise();
     }
+
+    return depTrees;
+  }
+
+  private flattenLockfile(lockfile: PackageLock): DepMap {
+    const depMap: DepMap = {};
+
+    const flattenLockfileRec = (lockfileDeps: PackageLockDeps, path: string[]) => {
+      for (const [depName, dep] of _.entries(lockfileDeps)) {
+        const depNode: DepMapItem = {
+          depType: dep.dev ? DepType.dev : DepType.prod,
+          dependencies: {},
+          name: depName,
+          requires: [],
+          version: dep.version,
+        };
+
+        if (dep.requires) {
+          depNode.requires = Object.keys(dep.requires);
+        }
+
+        const depKey: string[] = [...path, depName];
```

Comment:
would call this `depPath`/`depPathArray`, and `depKey` would be the `join`-ed string